crates/astria-sequencer/src/app/vote_extension.rs

```diff
@@ -0,0 +1,649 @@
+    ensure,
+    Context as _,
+};
+use astria_core::{
+    crypto::Signature,
+    generated::astria_vendored::slinky::{
+        abci::v1::OracleVoteExtension as RawOracleVoteExtension,
+        service::v1::{
+            oracle_client::OracleClient,
+            QueryPricesRequest,
+            QueryPricesResponse,
+        },
+    },
+    slinky::{
+        abci::v1::OracleVoteExtension,
+        oracle::v1::QuotePrice,
+        types::v1::CurrencyPair,
+    },
+};
+use indexmap::IndexMap;
+use prost::Message as _;
+use tendermint::{
+    abci,
+    abci::types::{
+        BlockSignatureInfo::Flag,
+        CommitInfo,
+        ExtendedCommitInfo,
+    },
+};
+use tendermint_proto::google::protobuf::Timestamp;
+use tonic::transport::Channel;
+use tracing::{
+    debug,
+    trace,
+};
+
+use crate::{
+    address::StateReadExt as _,
+    authority::StateReadExt as _,
+    slinky::oracle::{
+        currency_pair_strategy::DefaultCurrencyPairStrategy,
+        state_ext::StateWriteExt,
+    },
+    state_ext::StateReadExt,
+};
+
+// https://github.com/skip-mev/slinky/blob/793b2e874d6e720bd288e82e782502e41cf06f8c/abci/types/constants.go#L6
+const MAXIMUM_PRICE_BYTE_LEN: usize = 33;
+
+pub(crate) struct Handler {
+    // gRPC client for the slinky oracle sidecar.
+    oracle_client: Option<OracleClient<Channel>>,
+}
+
+impl Handler {
+    pub(crate) fn new(oracle_client: Option<OracleClient<Channel>>) -> Self {
+        Self {
+            oracle_client,
+        }
+    }
+
+    pub(crate) async fn extend_vote<S: StateReadExt>(
+        &mut self,
+        state: &S,
+    ) -> anyhow::Result<abci::response::ExtendVote> {
+        let Some(oracle_client) = self.oracle_client.as_mut() else {
+            // we allow validators to *not* use the oracle sidecar currently,
+            // so this will get converted to an empty vote extension when bubbled up.
+            //
+            // however, if >1/3 of validators are not using the oracle, the prices will not update.
+            bail!("oracle client not set")
+        };
+
+        // if we fail to get prices within the timeout duration, we will return an empty vote
+        // extension to ensure liveness.
+        let prices = match oracle_client.prices(QueryPricesRequest {}).await {
+            Ok(prices) => prices.into_inner(),
+            Err(e) => {
+                bail!("failed to get prices from oracle sidecar: {e}",);
+            }
+        };
+
+        tracing::debug!(
+            prices_count = prices.prices.len(),
+            "got prices from oracle sidecar; transforming prices"
+        );
+
+        let oracle_vote_extension = transform_oracle_service_prices(state, prices)
+            .await
+            .context("failed to transform oracle service prices")?;
+
+        Ok(abci::response::ExtendVote {
+            vote_extension: oracle_vote_extension.into_raw().encode_to_vec().into(),
+        })
+    }
+
+    pub(crate) async fn verify_vote_extension<S: StateReadExt>(
```

Comment:
Is it intended that if the oracle is disabled, that this function is still called? If the oracle is disabled, `NO_ORACLE=true`, I would expect that a sequencer node completely abstains from participating in vote extensions.

I think it might be clearer if we separated the config in 1. participates in vote extensions, 2. only verifies vote extensions, 3. ignores vote extensions?

Reply:
if the oracle sidecar is disabled, this function should still be called as it's part of consensus, and validator nodes don't need the oracle sidecar to perform verification.

the cases are validator nodes which run the oracle sidecar, those that don't, and full nodes. all validator nodes need to perform verification for consensus reasons, however it's optional as to whether they want to run the sidecar which publishes prices to the network as vote extensions. validators shouldn't ignore vote extension verification, as that could result in them accepting an invalid block. full nodes don't care about vote extensions, to them the VEs are just another transaction in a block (the one the proposer now places first).